### ISSUE TITLE
[Question] Can't use tab for both snippet jumping and popup jump completion

### ISSUE THREAD

**Issue by hungpham3112** [neoclide/coc.nvim#3745]
In LunarVim, they can use `<Tab>` for both navigate next item and snippet expand. I put the script based on coc-nvim document to my vimrc.

```
Map <tab> for trigger completion, completion confirm, snippet expand and jump
like VSCode. >

	inoremap <silent><expr> <TAB>
	  \ pumvisible() ? coc#_select_confirm() :
	  \ coc#expandableOrJumpable() ?
	  \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
	  \ <SID>check_back_space() ? "\<TAB>" :
	  \ coc#refresh()

	function! s:check_back_space() abort
	  let col = col('.') - 1
	  return !col || getline('.')[col - 1]  =~# '\s'
	endfunction

	let g:coc_snippet_next = '<tab>'
```

I only can use `<tab>` for snippet jumping, popup navigation doesn't work, it always pick the first item in list.

https://user-images.githubusercontent.com/75968004/161657971-fa8af36c-7e75-4522-a3c9-e47795eeb360.mp4

Config:

```
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"                             Plugin: Coc-nvim                               "
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"Completion settings
"use <tab> for trigger completion and navigate to the next complete item
"function! s:check_back_space() abort
    "let col = col('.') - 1
    "return !col || getline('.')[col - 1]  =~# '\s'
"endfunction

"inoremap <silent><expr> <TAB>
    "\ pumvisible() ? coc#_select_confirm() :
    "\ coc#expandableOrJumpable() ?
    "\ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
"\ <SID>check_back_space() ? "\<TAB>" :
"\ coc#refresh()

"inoremap <silent><expr> <TAB>
"\ pumvisible() ? "\<C-n>" :
"\ <SID>check_back_space() ? "\<TAB>" :
"\ coc#refresh()

""""""""""""""""""""""""""""""""""""""""""""""""""""""
inoremap <silent><expr> <TAB>
  \ pumvisible() ? coc#_select_confirm() :
  \ coc#expandableOrJumpable() ?
  \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
  \ <SID>check_back_space() ? "\<TAB>" :
  \ coc#refresh()

function! s:check_back_space() abort
  let col = col('.') - 1
  return !col || getline('.')[col - 1]  =~# '\s'
endfunction

let g:coc_snippet_next = '<tab>'
autocmd
    94d13f69-3e60-4a7a-bd6b-2c402bae1ad5
        0d6059ae-d149-44a6-a91a-c164ee5e3ce5
augroup AU_NAME
    autocmd!
    autocmd BufRead,BufNewFile *.ext,*.ext3|<buffer[=N]>        
augroup END
""""""""""""""""""""""""""""""""""""""""""""""""""""""
"Select the first completion item and confirm the completion when no item has been selected:
inoremap <silent><expr> <cr> pumvisible() ? coc#_select_confirm()
    \: "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"

"let g:coc_snippet_next = '<M-l>'
"let g:coc_snippet_prev = '<M-h>'

if has('nvim-0.4.0') || has('patch-8.2.0750')
  nnoremap <silent><nowait><expr> <C-d> coc#float#has_scroll() ? coc#float#scroll(1) : "\<C-f>"
  nnoremap <silent><nowait><expr> <C-u> coc#float#has_scroll() ? coc#float#scroll(0) : "\<C-b>"
  inoremap <silent><nowait><expr> <C-d> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(1)\<cr>" : "\<Right>"
  inoremap <silent><nowait><expr> <C-u> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(0)\<cr>" : "\<Left>"
  vnoremap <silent><nowait><expr> <C-d> coc#float#has_scroll() ? coc#float#scroll(1) : "\<C-f>"
  vnoremap <silent><nowait><expr> <C-u> coc#float#has_scroll() ? coc#float#scroll(0) : "\<C-b>"
endif

nnoremap <silent> K :call <SID>show_documentation()<CR>

function! s:show_documentation()
    if (index(['vim','help'], &filetype) >= 0)
        execute 'h '.expand('<cword>')
    elseif (coc#rpc#ready())
        call CocActionAsync('doHover')
    else
        execute '!' . &keywordprg . " " . expand('<cword>')
    endif
endfunction

if exists('g:did_coc_loaded') 
    autocmd CursorHold * silent call CocActionAsync('highlight')
endif

let g:coc_global_extensions = ['coc-json',
            \'coc-vimlsp',
            \'coc-snippets',
            \'coc-powershell',
            \'coc-markdownlint',]

```

**Comment by chemzqm** [comment]
Use `<C-n>` and `<C-p>` to navigate pum list, change `pumvisible() ? coc#_select_confirm() :` to fit your need.

**Comment by hungpham3112** [comment]
> It's a bad idea since you may want to jump next placeholder and you have pum visible as the same time.

Did you try Lunarvim? It's not a bad idea, it's ergonomic way because your hand doesn't have to move.

**Comment by chemzqm** [comment]
No, it's up to you.

**Comment by hungpham3112** [comment]
> In LunarVim, they can use `<Tab>` for both navigate next item and snippet expand. I put the script based on coc-nvim document to my vimrc.
> 
> ```
> Map <tab> for trigger completion, completion confirm, snippet expand and jump
> like VSCode. >
> 
> 	inoremap <silent><expr> <TAB>
> 	  \ pumvisible() ? coc#_select_confirm() :
> 	  \ coc#expandableOrJumpable() ?
> 	  \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
> 	  \ <SID>check_back_space() ? "\<TAB>" :
> 	  \ coc#refresh()
> 
> 	function! s:check_back_space() abort
> 	  let col = col('.') - 1
> 	  return !col || getline('.')[col - 1]  =~# '\s'
> 	endfunction
> 
> 	let g:coc_snippet_next = '<tab>'
> ```
> 
> I only can use `<tab>` for snippet jumping, popup navigation doesn't work, it always pick the first item in list.
> 
>  bandicam.2022-04-05.08-03-43-762.mp4 
> Config:
> 
> ```
> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
> "                             Plugin: Coc-nvim                               "
> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
> "Completion settings
> "use <tab> for trigger completion and navigate to the next complete item
> "function! s:check_back_space() abort
>     "let col = col('.') - 1
>     "return !col || getline('.')[col - 1]  =~# '\s'
> "endfunction
> 
> "inoremap <silent><expr> <TAB>
>     "\ pumvisible() ? coc#_select_confirm() :
>     "\ coc#expandableOrJumpable() ?
>     "\ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
> "\ <SID>check_back_space() ? "\<TAB>" :
> "\ coc#refresh()
> 
> "inoremap <silent><expr> <TAB>
> "\ pumvisible() ? "\<C-n>" :
> "\ <SID>check_back_space() ? "\<TAB>" :
> "\ coc#refresh()
> 
> """"""""""""""""""""""""""""""""""""""""""""""""""""""
> inoremap <silent><expr> <TAB>
>   \ pumvisible() ? coc#_select_confirm() :
>   \ coc#expandableOrJumpable() ?
>   \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
>   \ <SID>check_back_space() ? "\<TAB>" :
>   \ coc#refresh()
> 
> function! s:check_back_space() abort
>   let col = col('.') - 1
>   return !col || getline('.')[col - 1]  =~# '\s'
> endfunction
> 
> let g:coc_snippet_next = '<tab>'
> autocmd
>     94d13f69-3e60-4a7a-bd6b-2c402bae1ad5
>         0d6059ae-d149-44a6-a91a-c164ee5e3ce5
> augroup AU_NAME
>     autocmd!
>     autocmd BufRead,BufNewFile *.ext,*.ext3|<buffer[=N]>        
> augroup END
> """"""""""""""""""""""""""""""""""""""""""""""""""""""
> "Select the first completion item and confirm the completion when no item has been selected:
> inoremap <silent><expr> <cr> pumvisible() ? coc#_select_confirm()
>     \: "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"
> 
> "let g:coc_snippet_next = '<M-l>'
> "let g:coc_snippet_prev = '<M-h>'
> 
> if has('nvim-0.4.0') || has('patch-8.2.0750')
>   nnoremap <silent><nowait><expr> <C-d> coc#float#has_scroll() ? coc#float#scroll(1) : "\<C-f>"
>   nnoremap <silent><nowait><expr> <C-u> coc#float#has_scroll() ? coc#float#scroll(0) : "\<C-b>"
>   inoremap <silent><nowait><expr> <C-d> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(1)\<cr>" : "\<Right>"
>   inoremap <silent><nowait><expr> <C-u> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(0)\<cr>" : "\<Left>"
>   vnoremap <silent><nowait><expr> <C-d> coc#float#has_scroll() ? coc#float#scroll(1) : "\<C-f>"
>   vnoremap <silent><nowait><expr> <C-u> coc#float#has_scroll() ? coc#float#scroll(0) : "\<C-b>"
> endif
> 
> nnoremap <silent> K :call <SID>show_documentation()<CR>
> 
> function! s:show_documentation()
>     if (index(['vim','help'], &filetype) >= 0)
>         execute 'h '.expand('<cword>')
>     elseif (coc#rpc#ready())
>         call CocActionAsync('doHover')
>     else
>         execute '!' . &keywordprg . " " . expand('<cword>')
>     endif
> endfunction
> 
> if exists('g:did_coc_loaded') 
>     autocmd CursorHold * silent call CocActionAsync('highlight')
> endif
> 
> let g:coc_global_extensions = ['coc-json',
>             \'coc-vimlsp',
>             \'coc-snippets',
>             \'coc-powershell',
>             \'coc-markdownlint',]
> ```

So why in coc.nvim document said: 

>Map <tab> for trigger completion, completion confirm, snippet expand and jump
like VSCode. >

It's said both trigger completion and snippet expand?

**Comment by chemzqm** [comment]
You can use that example code or not, so what is your problem?

**Comment by hungpham3112** [comment]
> You can use that example code or not, so what is your problem?

I want to use tab for both snippet jump to placeholder and navigate the popup. I don't know how to config. The snippet in coc-nvim document doesn't work to me.

**Comment by chemzqm** [comment]
Learn viml, the plugin already provided what you need.

**Comment by hungpham3112** [comment]
Fixed at: [#3766 (comment)](https://github.com/neoclide/coc.nvim/issues/3766#issuecomment-1102933729)
